libp2p/switch.nim

```diff
@@ -417,30 +426,60 @@ proc mount*[T: LPProtocol](s: Switch, proto: T, matcher: Matcher = nil) {.gcsafe
   s.ms.addHandler(proto.codecs, proto, matcher)
 
 proc accept(s: Switch, transport: Transport) {.async.} = # noraises
-  ## transport's accept loop
+  ## switch accept loop, ran for every transport
   ##
 
+  let upgrades = AsyncSemaphore.init(ConcurrentUpgrades)
   while transport.running:
     var conn: Connection
     try:
       debug "About to accept incoming connection"
-      conn = await transport.accept()
-      if not isNil(conn):
-        debug "Accepted an incoming connection", conn
-        asyncSpawn s.upgradeIncoming(conn) # perform upgrade on incoming connection
-      else:
+      # remember to always release the slot when
+      # the upgrade succeeds or fails, this is
+      # currently done by the `upgradeMonitor`
+      await upgrades.acquire()        # first wait for an upgrade slot to become available
+      conn = await transport.accept() # next attempt to get a connection
+      if isNil(conn):
         # A nil connection means that we might have hit a
         # file-handle limit (or another non-fatal error),
         # we can get one on the next try, but we should
         # be careful to not end up in a thigh loop that
         # will starve the main event loop, thus we sleep
         # here before retrying.
+        trace "Unable to get a connection, sleeping"
         await sleepAsync(100.millis) # TODO: should be configurable?
+        upgrades.release()
+        continue
+
+      debug "Accepted an incoming connection", conn
+      proc upgradeMonitor() {.async.} =
+        ## monitor connection for upgrades
+        ##
+        try:
+          # Since we don't control the flow of the
+          # upgrade, this timeout guarantees that a
+          # "hanged" remote doesn't hold the upgrade
+          # forever
+          await conn.upgraded.wait(30.seconds) # wait for connection to be upgraded
+          trace "Connection upgrade succeeded"
+        except CatchableError as exc:
+          if not isNil(conn): # for some reason, this can be nil
```

Comment:
This `nil` is puzzling me and it also happens almost immediately so quite reproducible, and yet there isn't anything that would `nillify` the connection explicitly and the closure scope should still be holding onto it - possible GC bug?

Reply:
eeh... right, nim has some pretty weird closure rules - wonder for example if a new instance is allocated on every iteration - @zah?

Reply:
the more reliable way to write things is to not use a closure and pass the variable in explicitly, then you don't get hit by gotchas